flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.flink.shaded.curator4.org.apache.curator.framework.recipes.cache.ChildData;
+import org.apache.flink.shaded.curator4.org.apache.curator.framework.recipes.cache.NodeCache;
+import org.apache.flink.shaded.curator4.org.apache.curator.framework.recipes.cache.NodeCacheListener;
+import org.apache.flink.shaded.curator4.org.apache.curator.framework.state.ConnectionState;
+import org.apache.flink.shaded.curator4.org.apache.curator.framework.state.ConnectionStateListener;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.UUID;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The counterpart to the {@link ZooKeeperLeaderElectionDriver}.
+ * {@link LeaderRetrievalService} implementation for Zookeeper. It retrieves the current leader which has
+ * been elected by the {@link ZooKeeperLeaderElectionDriver}.
+ * The leader address as well as the current leader session ID is retrieved from ZooKeeper.
+ */
+public class ZooKeeperLeaderRetrievalDriver implements LeaderRetrievalDriver, NodeCacheListener, UnhandledErrorListener {
+	private static final Logger LOG = LoggerFactory.getLogger(ZooKeeperLeaderRetrievalDriver.class);
+
+	/** Connection to the used ZooKeeper quorum. */
+	private final CuratorFramework client;
+
+	/** Curator recipe to watch changes of a specific ZooKeeper node. */
+	private final NodeCache cache;
+
+	private final String retrievalPath;
+
+	private final ConnectionStateListener connectionStateListener = (client, newState) -> handleStateChange(newState);
+
+	private final LeaderRetrievalEventHandler leaderRetrievalEventHandler;
+
+	private volatile boolean running;
+
+	/**
+	 * Creates a leader retrieval service which uses ZooKeeper to retrieve the leader information.
+	 *
+	 * @param client Client which constitutes the connection to the ZooKeeper quorum
+	 * @param retrievalPath Path of the ZooKeeper node which contains the leader information
+	 * @param leaderRetrievalEventHandler handler to notify the leader changes.
+	 */
+	public ZooKeeperLeaderRetrievalDriver(
+			CuratorFramework client,
+			String retrievalPath,
+			LeaderRetrievalEventHandler leaderRetrievalEventHandler) throws Exception {
+		this.client = checkNotNull(client, "CuratorFramework client");
+		this.cache = new NodeCache(client, retrievalPath);
+		this.retrievalPath = checkNotNull(retrievalPath);
+
+		this.leaderRetrievalEventHandler = checkNotNull(leaderRetrievalEventHandler);
+
+		client.getUnhandledErrorListenable().addListener(this);
+		cache.getListenable().addListener(this);
+		cache.start();
+
+		client.getConnectionStateListenable().addListener(connectionStateListener);
+
+		running = true;
+	}
+
+	@Override
+	public void close() throws Exception {
+		if (!running) {
+			return;
+		}
+
+		running = false;
```

Comment:
I'd suggest to introduce a `lock` to guard access to running (also at other places where we access this field). That way we will make sure that after `close` is called that we won't see any calls coming from the `ZooKeeper` threads.

Reply:
At the very beginning, I think the `volatile` is enough. But after a careful consideration, you are right. We need a lock to guard access of `running`.

Reply:
I think we just need to make sure that we don't introduce a deadlock by requesting the inner and outer lock in a problematic order. In order to ensure this we should not call the event handler under the inner lock.